Created DiffieHellman.java and MonoAlphabetic.java

riti2601 opened the pull request:
<!--
Thank you for your contribution!
In order to reduce the number of notifications sent to the maintainers, please:
- create your PR as draft, cf. https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests#draft-pull-requests,
- make sure that all of the CI checks pass,
- mark your PR as ready for review, cf. https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request#marking-a-pull-request-as-ready-for-review
-->
Description

This PR implements the Diffie-Hellman Key Exchange and MonoAlphabetic Cipher algorithms. 

Key Changes:

1. Diffie-Hellman Key Exchange:
   - Added `DiffieHellman.java` under the `ciphers` directory.
   - The implementation allows two users (Person A and Person B) to securely generate a shared secret key over an insecure channel using large prime numbers and modular arithmetic.
   - The code prompts the users to input their secret numbers and performs the necessary calculations to exchange and compute the shared secret keys.

2. MonoAlphabetic Cipher:
   - Added `MonoAlphabetic.java` under the `ciphers` directory.
   - The implementation allows the user to either **encrypt** or **decrypt** data using a predefined substitution cipher, where each letter of the alphabet is replaced with a corresponding letter.
   - The user is prompted to enter the data, and the cipher is applied to return either the encrypted or decrypted result.

Both algorithms are utility classes, with private constructors to prevent instantiation and static methods to handle their respective functionalities.


<!-- For completed items, change [ ] to [x] -->

- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Java/blob/master/CONTRIBUTING.md).
- [x] This pull request is all my own work -- I have not plagiarized it.
- [x] All filenames are in PascalCase.
- [x] All functions and variable names follow Java naming conventions.
- [x] All new algorithms have a URL in their comments that points to Wikipedia or other similar explanations.
- [x] All new code is formatted with `clang-format -i --style=file path/to/your/file.java`

vil02 reviewed (changes requested):
Please remove the `main` methods and add proper junit tests. Prefer `ParameterizedTest`s.

riti2601 commented:
I am currently unable to build my code due to the outdated pom.xml file. The dependencies required for the JUnit tests are missing, leading to compilation errors. To proceed with testing and building my changes effectively, I recommend updating the pom.xml file to include the necessary JUnit dependencies.

Please let me know if there's a preferred way to address this issue or if I should modify the code to remove the dependencies temporarily.

Thank you!

vil02 commented:
Did you try to place the tests in the [`.../src/test/java/com/thealgorithms`](https://github.com/TheAlgorithms/Java/tree/842ff5294f5e97dfd1e4c09f9d9e01a89ac16e32/src/test/java/com/thealgorithms) directory? cf. #5501.

riti2601 commented:
No, I'll try it. 
Thank you

riti2601 commented:
Hi @vil02,
I've addressed all the requested changes and updated the pull request accordingly. Could you please review the latest changes at your convenience and let me know if anything else is needed?
Thank you!

vil02 commented on src/test/java/com/thealgorithms/ciphers/DiffieHellmanTest.java:
```diff
@@ -0,0 +1,38 @@
+        assertEquals(expected, dh.calculateSharedSecret(otherPublicValue)); // Call instance method
+    }
+
+    // Provide test data for public key calculation
+    private static Stream<Arguments> providePublicKeyData() {
+        return Stream.of(Arguments.of(BigInteger.valueOf(5), BigInteger.valueOf(6), BigInteger.valueOf(23), BigInteger.valueOf(8)), Arguments.of(BigInteger.valueOf(2), BigInteger.valueOf(5), BigInteger.valueOf(13), BigInteger.valueOf(6)));
+    }
+
+    // Provide test data for shared secret calculation
+    private static Stream<Arguments> provideSharedSecretData() {
+        return Stream.of(Arguments.of(BigInteger.valueOf(5), BigInteger.valueOf(6), BigInteger.valueOf(23), BigInteger.valueOf(8), BigInteger.valueOf(13)), Arguments.of(BigInteger.valueOf(2), BigInteger.valueOf(5), BigInteger.valueOf(13), BigInteger.valueOf(6), BigInteger.valueOf(2)));
+    }
```
Similarly here.

vil02 commented on src/main/java/com/thealgorithms/ciphers/MonoAlphabetic.java:
```diff
@@ -0,0 +1,44 @@
+
+public final class MonoAlphabetic {
+
+    // Private constructor to prevent instantiation of utility class
```
```suggestion
```

vil02 commented on src/main/java/com/thealgorithms/ciphers/MonoAlphabetic.java:
```diff
@@ -0,0 +1,44 @@
+    // Encryption method
+    public static String encrypt(String data, String key) {
+        StringBuilder sb = new StringBuilder();
+        data = data.toUpperCase();
```
I would suggest to throw an exception if the input has lower case letters.

riti2601 replied:
Added the exception accordingly.

vil02 commented on src/main/java/com/thealgorithms/ciphers/MonoAlphabetic.java:
```diff
@@ -0,0 +1,44 @@
+
+        for (char c : data.toCharArray()) {
+            if (c >= 'A' && c <= 'Z') {
+                int idx = c - 'A'; // Index in alphabet
```
Consider adding two methods: `charToPos` and `posToChar`.

riti2601 replied:
Added the methods charToPos and posToChar.

vil02 commented on src/main/java/com/thealgorithms/ciphers/DiffieHellman.java:
```diff
@@ -0,0 +1,36 @@
+        if (otherPublicValue == null || otherPublicValue.signum() <= 0) {
+            throw new IllegalArgumentException("Other public value must be non-null and positive.");
+        }
+        // Returns b^x mod p or a^y mod p
```
?

riti2601 replied:
These values are used to compute a shared secret without directly transmitting it, ensuring secure communication.

vil02 commented on src/main/java/com/thealgorithms/ciphers/DiffieHellman.java:
```diff
@@ -0,0 +1,36 @@
+
+    // Method to calculate public value (g^x mod p)
+    public BigInteger calculatePublicValue() {
+        // Returns g^x mod p
```
What is `g`, `x` and `p`?

riti2601 replied:
In the Diffie-Hellman key exchange:
1. g represents the base (also called the generator), which is a publicly agreed-upon number used by both parties to perform the key exchange.
2. p is a large prime number, another public value that ensures the security of the key exchange.
3. x is the private secret key chosen by one of the parties. This value is kept secret and used to compute the public key, which is shared with the other party.
Together, these values are used to compute a shared secret without directly transmitting it, ensuring secure communication.

riti2601 reviewed (commented):
Hi @vil02,
I have made the necessary changes according to your suggestions.
Thank you for your valuable feedback!

siriak commented:
@vil02 please review

riti2601 commented:
Hi @vil02, kindly review the changed code.

riti2601 commented:
Hello @vil02
I am participating in Hacktoberfest'25 and request that you kindly check the PR for further merging. 
Thank you
